rep-0145.rst

```diff
@@ -0,0 +1,207 @@
+REP: 145
+Title: Conventions for IMU Sensor Drivers
+Author: Paul Bovbel <pbovbel@clearpathrobotics.com>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 02-Feb-2015
+Post-History: 02-Feb-2015
+
+
+Abstract
+========
+
+This REP defines common parameters, topics, namespaces, and data processing conventions for drivers of Inertial Measurement Unit (IMU) sensors. This includes accelerometers, gyroscopes, magnetomers, and any combination thereof.
+
+Motivation
+==========
+
+This REP seeks to standardize an interface for IMU drivers in the ROS ecosystem, providing for IMU-type sensors with a variety of capabilities and conventions. By formalizing a common approach based on existing driver implementations, this REP serves as a guideline for new driver development, while minimizing the changes required to non-conforming implementations. This REP discourages the in-driver manipulation of sensor data, by specifying the conditions where sensor data will be properly handled by ROS transform and coordinate specifications. The REP also specifies fallback approaches to correcting sensor data that does not conform to ROS conventions.
+
+Specification
+=============
+
+Common Parameters
+-----------------
+
+IMU driver implementations should read as many of these parameters as are relevant.
+
+* `~port` (`string`)
+
+  - Represents the serial port used to connect to the device.
+
+* `~baud` (`int`)
+
+  - Represents the baud rate of the serial link.
+
+* `~frame_id` (`string`, default: `imu_link` or `imu_link_ned`)
+
+  - The frame ID to set in outgoing messages.
+
+* `~autocalibrate` (`bool`)
+
+  - Perform a calibration routine on node startup.
+
+* `~linear_acceleration_stddev` (`double`)
+
+  - Square root of the linear_acceleration_covariance diagonal elements in m/s^2. Overrides any values reported by the sensor.
+
+* `~angular_velocity_stddev` (`double`)
+
+  - Square root of the angular_velocity_covariance diagonal elements in rad/s. Overrides any values reported by the sensor.
+
+* `~magnetic_field_stddev` (`double`)
+
+  - Square root of the magnetic_field_covariance diagonal elements in Tesla. Overrides any values reported by the sensor.
+
+* `~orientation_stddev` (`double`)
+
+  - Square root of the orientation_covariance diagonal elements in rad. Overrides any values reported by the sensor.
+
+Frame Conventions
+-----------------
+
+An IMU device may measure data with respect to two frames, specified by the manufacturer:
+
+* The **body frame** represents the internal device axes. These may be found either in the device specification documents, and sometimes printed directly on the device body. This frame is fixed to the device orientation.
+
+* If an orientation estimate is provided (see `Data Reporting`_), the **world frame** represents the external reference frame for the device. This frame is fixed in one of two orientations, with the relevant conventions from REP 103 [1]_ :
```

Comment:
It should be clarified whether the north axis of the world frame points to magnetic or geographic north. The difference between both is the [declination or variation](http://en.wikipedia.org/wiki/Magnetic_declination) and can be up to 180 degrees.

Most IMUs do probably not consider declination in their yaw output, so either the world frame refers to magnetic north or the IMU driver should already provide a parameter to correct the magnetic heading by the local declination. In the first case I would vote to use a different frame name (e.g. **magnetic reference frame**) because orientations in this frame should not be directly compared with geographical orientations and directions (e.g. from GPS).